src/turret.cpp

```diff
@@ -0,0 +1,382 @@
+        }
+
+        tripoint pos = global_part_pos3( *e );
+        const int rng = e->base.gun_range();
+
+        int res = 0;
+        res = std::max( res, rl_dist( g->u.pos(), { pos.x + rng, pos.y, pos.z } ) );
+        res = std::max( res, rl_dist( g->u.pos(), { pos.x - rng, pos.y, pos.z } ) );
+        res = std::max( res, rl_dist( g->u.pos(), { pos.x, pos.y + rng, pos.z } ) );
+        res = std::max( res, rl_dist( g->u.pos(), { pos.x, pos.y - rng, pos.z } ) );
+        return std::max( lhs, res );
+    } );
+
+    // fake gun item to aim
+    item pointer( "vehicle_pointer" );
+    pointer.set_curammo( "pointer_fake_ammo" );
+    pointer.ammo_data()->ammo->range = range;
+
+    tripoint pos = g->u.pos();
+    std::vector<tripoint> trajectory;
+
+    if( opts.empty() ) {
+        add_msg( m_warning, _( "Can't aim turrets: all turrets are offline" ) );
+    } else {
+        trajectory = g->pl_target_ui( pos, range, &pointer, TARGET_MODE_TURRET );
+        g->draw_ter();
+    }
+
+    if( !trajectory.empty() ) {
+        // set target for any turrets in range
+        for( auto e : turrets( trajectory.back() ) ) {
+            e->target.second = trajectory.back();
+        }
+        ///\EFFECT_INT speeds up aiming of vehicle turrets
+        g->u.moves = std::min( 0, g->u.moves - 100 + ( 5 * g->u.int_cur ) );
+    }
+
+    for( auto e : opts ) {
+        turret_unload( *e );
+    }
+
+    return !trajectory.empty();
+}
+
+void vehicle::turret_reload( vehicle_part &pt )
+{
+    item &gun = pt.base;
+    if( !gun.is_gun() || gun.ammo_type() == "NULL" ) {
+        return;
+    }
+
+    // try to reload using stored fuel from onboard vehicle tanks
+    for( const auto &e : fuels_left() ) {
+        const itype *fuel = item::find_type( e.first );
+        if( !fuel->ammo ) {
+            continue;
+        }
+
+        if( ( gun.ammo_current() == "null" && fuel->ammo->type == gun.ammo_type() ) ||
+            ( gun.ammo_current() == e.first ) ) {
+
+            int qty = gun.ammo_remaining();
+            gun.ammo_set( e.first, e.second );
+
+            drain( e.first, gun.ammo_remaining() - qty );
+
+            return;
+        }
+    }
+
+    // otherwise try to reload from turret cargo space
+    for( auto it = pt.items.begin(); it != pt.items.end(); ++it ) {
+        if( it->is_ammo() &&
+            ( ( gun.ammo_current() == "null" && it->ammo_type() == gun.ammo_type() ) ||
+              ( gun.ammo_current() == it->typeId() ) ) ) {
+
+            int qty = gun.ammo_remaining();
+            gun.ammo_set( it->typeId(), it->charges );
+
+            it->charges -= gun.ammo_remaining() - qty;
+
+            if( it->charges <= 0 ) {
+                pt.items.erase( it );
+            }
+
+            return;
+        }
+    }
+}
+
+void vehicle::turret_unload( vehicle_part &pt )
+{
+    item &gun = pt.base;
+    if( !gun.is_gun() || gun.ammo_type() == "NULL" ||
+        gun.ammo_current() == "null" || gun.ammo_remaining() == 0 ) {
```

Comment:
Not necessary here, just noting:

`ammo_type() == "NULL"` and `ammo_current() == "null"` are repeated a lot in the code and are rather ugly.

It would be much better if they looked like `ammo_type().is_null()` and `ammo_current().is_null()`.

The fact that they use different capitalization adds to the confusion and has resulted in at least one bug, as far as I recall.